proposals/0030-markdoc.md

```diff
@@ -0,0 +1,327 @@
+```
+
+See [the detailed API section](#api) for information on supported Markdoc configuration, shared components config, and manual rendering.
+
+# Background & Motivation
+
+We've received multiple user requests for Markdoc support since its public launch. In fact, we've seen an early community project, [Blogster](https://github.com/flexdinesh/blogster), that brings Markdoc to robust project themes. This includes an [Astro Markdoc renderer](https://github.com/flexdinesh/blogster/tree/main/packages/astro-markdoc-renderer) that may inspire a first party integration.
+
+Markdoc is also designed to solve existing limitations of MDX in Astro:
+
+1. **Performance suffers at scale.** Unlike plain Markdown that outputs a string of HTML, MDX outputs a JavaScript module of JSX components. This requires Astro and Babel preprocessing for even the simplest MDX documents. Notably, this required a bump to our maximum memory usage when building / deploying docs.astro.build after migrating to MDX.
+2. **Your content is tied to your UI.** MDX can import styles and components directly, which _is_ convenient from a developer standpoint. However, this causes issues when content needs to be reused in multiple contexts. A common example is RSS, where you may want a component-rich version from your blog and a simplified HTML output for your RSS feed.
+
+Markdoc is built to solve (2) by separating content from the components, styles, and assets you choose to render. You can use an Astro component renderer when using on your site, [Markdoc's own `html` renderer](https://markdoc.dev/docs/render#html) for RSS, and even write your own renderer to traverse Markdoc pages yourself. (1) Is something we're excited to test, requiring a thorough performance benchmark.
+
+The content collections API was built generically to support this future, choosing format-agnostic naming like `data` instead of `frontmatter` and `body` instead of `rawContent`. Because of this, introducing new authoring formats is possible without breaking changes.
+
+# Goals
+
+- **Create an `@astrojs/markdoc` integration** that adds `.mdoc` support to content collections.
+- **Support Astro components and server-rendered UI components** (React, Vue, Svelte, etc) within Markdoc files. Note this leaves client-rendered UI components out-of-scope (see non-goals).
+- **Benchmark Markdoc performance against Markdown and MDX** at 1k and 10k documents. This addresses problem (1) from the previous section.
+
+# Non-goals
+
+- Non-goal: **ESM import and `src/pages/` support** for Markdoc files. [See discussion](https://github.com/withastro/roadmap/discussions/478#discussioncomment-4975072) for context.
+- Out-of-scope: **Allowing the `.md` extension.** This would mean overriding Astro's `.md` renderer, which is tightly coupled to remark and your `markdown` configuration options today. We agree using `.md` for Markdoc is a common use case, and deserves a separate proposal to make Astro's Markdown rendering flexible.
```

Comment:
A little history into how I tried to parse markdoc in `*.md` extensions in Blogster and why it introduced a whole lot of unintended complexity and why I ended up not doing in.

This was one of the dilemmas I had when building Blogster. The original motivation was that I didn't want users to choose b/w markdoc and markdown. I wanted markdoc to be an extension of markdown in a way that when users don't write the special markdoc component syntax (`{% custom-component attr="val" %}`) in their `*.md` files, the content would be parsed with Astro's markdown parser during build and if markdoc specific syntax is detected, the markdoc parser would be used.

Few reasons why I thought this would be good for users:
- Aligns well with Astro's philosophy of default to simple and extend progressively
- Users don't have to decide b/w markdown vs markdoc. All files are `*.md` files and markdoc is just an extension of markdown
- `*.md` is a standard extension and UI tools like GitHub web UI for PRs/etc has a preview mode for markdown files. So markdown compatible syntax in markdoc files could still benefit from the syntax highlighting and preview options in such tools

However when I tried writing a custom build plugin to solve this I realised how much complex it was to smartly switch parsers when both markdoc and markdown content have the same entension `*.md`. I dropped the idea and ended up building the parser config as a utility here: [Blogster utility to read and parse *.md files with markdoc parser](https://github.com/flexdinesh/blogster/blob/386a10a70d3e2f7dd92bf887fda7a62896cfd96f/templates/bubblegum/src/lib/markdoc/read.ts#L99) instead of building it as an Astro integration.

Reply:
Well said! I agree that `*.md` support would be a nicety that aligns with Markdoc's own documentation. Still, [I'll echo Erika's perspective on this](https://github.com/withastro/roadmap/discussions/503#discussioncomment-5123556): Pointing multiple formats to a single extension makes language tooling difficult. We've noticed that Markdoc is _not_ a strict superset of Markdown (it namely does not allow plain HTML in Markdown), so this can lead to a muddled experience as different extensions compete. I also think that flipping a file extension from `.md` to `.mdoc` is still a simple story for incremental adoption.

I do like Blogster's approach of making `.md` support a separate plug-and-play experience though. Would you be open to making a discussion post on this if it's a feature we want?